homeassistant/components/mqtt/client.py

```diff
@@ -759,17 +759,25 @@ def _mqtt_handle_message(self, msg: mqtt.MQTTMessage) -> None:
                         subscription.job,
                     )
                     continue
-            self.hass.async_run_hass_job(
-                subscription.job,
-                ReceiveMessage(
-                    msg.topic,
-                    payload,
-                    msg.qos,
-                    msg.retain,
-                    subscription.topic,
-                    timestamp,
-                ),
-            )
+
+            # Skip messages for non-wildcard topics where RETAIN=1, and we have
+            # already received a RETAIN=1 message.
+            if not _is_simple_match(subscription.topic) or not (
+                msg.retain and subscription.initialized
+            ):
+                if msg.retain:
+                    subscription.initialized = True
+                self.hass.async_run_hass_job(
+                    subscription.job,
+                    ReceiveMessage(
+                        msg.topic,
+                        payload,
+                        msg.qos,
+                        msg.retain,
+                        subscription.topic,
+                        timestamp,
+                    ),
+                )
```

Comment:
Let's do the debouncing of subscribe and potential filtering of received messages in separate PRs.

Reply:
Sure, both changes are means to improve the startup performance.
However, the means are completely different.

Reply:
I have a strong belief that, when you fix a bug, you fix ALL of a bug.  In one PR.  You don't do it piecemeal.   Whether it's due to a single issue, or a combination of several.  The PR is to fix the bug, NOT to fix one piece of the bug.  If that makes the PR bigger, *shrug*.  The idea is that, you should not be able to accept one without the other, because doing so is a partial fix at best and doesn't actually fix the bug in question.

As such, since everything here is fixing the same but, even though yes, it's a contribution of two different (complementing) issues, the fix belongs in one PR.  And so no, I'm not going to split them.  If you feel strongly about it, as I said, feel free to take over the branch and split it up.

Reply:
OK, let's close this PR then.